joblib/executor.py

```diff
@@ -8,64 +8,113 @@
 
-import random
 from ._memmapping_reducer import get_memmapping_reducers
-from ._memmapping_reducer import TemporaryResourcesManagerMixin
-from .externals.loky.reusable_executor import get_reusable_executor
+from ._memmapping_reducer import TemporaryResourcesManager
+from .externals.loky.reusable_executor import _ReusablePoolExecutor
 
 
 _executor_args = None
 
 
-def get_memmapping_executor(n_jobs, timeout=300, initializer=None, initargs=(),
-                            env=None, **backend_args):
-    """Factory for ReusableExecutor with automatic memmapping for large numpy
-    arrays.
-    """
-    global _executor_args
-    # Check if we can reuse the executor here instead of deferring the test to
-    # loky as the reducers are objects that changes at each call.
-    executor_args = backend_args.copy()
-    executor_args.update(env if env else {})
-    executor_args.update(dict(
-        timeout=timeout, initializer=initializer, initargs=initargs))
-    reuse = _executor_args is None or _executor_args == executor_args
-    _executor_args = executor_args
-
-    id_executor = random.randint(0, int(1e10))
-    job_reducers, result_reducers, temp_folder = get_memmapping_reducers(
-        id_executor, unlink_on_gc_collect=True, **backend_args)
-    _executor = get_reusable_executor(n_jobs, job_reducers=job_reducers,
-                                      result_reducers=result_reducers,
-                                      reuse=reuse, timeout=timeout,
-                                      initializer=initializer,
-                                      initargs=initargs, env=env)
-    # If executor doesn't have a _temp_folder, it means it is a new executor
-    # and the reducers have not been used. Else, the previous reducers are used
-    # and we should not change this attribute.
-    if not hasattr(_executor, "_temp_folder"):
-        _executor._temp_folder = temp_folder
-    return _executor
-
-
-class _TestingMemmappingExecutor(TemporaryResourcesManagerMixin):
+def get_memmapping_executor(n_jobs, **kwargs):
+    return MemmappingExecutor.get_memmapping_executor(n_jobs, **kwargs)
+
+
+class MemmappingExecutor(_ReusablePoolExecutor):
+
+    @classmethod
+    def get_memmapping_executor(cls, n_jobs, timeout=300, initializer=None,
+                                initargs=(), env=None, temp_folder=None,
+                                context_id=None, **backend_args):
+        """Factory for ReusableExecutor with automatic memmapping for large numpy
+        arrays.
+        """
+        global _executor_args
+        # Check if we can reuse the executor here instead of deferring the test
+        # to loky as the reducers are objects that changes at each call.
+        executor_args = backend_args.copy()
+        executor_args.update(env if env else {})
+        executor_args.update(dict(
+            timeout=timeout, initializer=initializer, initargs=initargs))
+        reuse = _executor_args is None or _executor_args == executor_args
```

Comment:
This check is not correct.
We should not reuse executor if `_executor_args is None`.
For example, this fails:
```python
from joblib.executor import _ReusablePoolExecutor, MemmappingExecutor

ex1 = _ReusablePoolExecutor.get_reusable_executor(2)  # inside another backend

ex2 = MemmappingExecutor.get_memmapping_executor(2)
assert isinstance(ex2, MemmappingExecutor)  # wrong type

ex3 = MemmappingExecutor.get_memmapping_executor(2, context_id=42)
# AttributeError: '_ReusablePoolExecutor' object has no attribute '_temp_folder_manager'
```
Correct line is:
`reuse = _executor_args == executor_args`

Reply:
Thanks for the report!!
Indeed, if `_executor_args` is `None`, we should never reuse as we did not created any `MemmappingExecutor` yet. I created a PR to track this issue #1052